### PR TITLE
#415 force signout

### PR DESCRIPTION
**Pull request by GD-Dheer**
This pr contains functionality to auto sign out a user. An alert will pop up after 30 seconds indicating how much time is left before session expires. Then redirects user to login

### CODE REVIEW

File: Client/src/Components/NavigationMenu.tsx [hunk 1]
```diff
@@ -96,6 +92,9 @@ export default function NavigationMenu() {
               </Grid>
             </Grid>
           </Toolbar>
+          {user.sessionExpiration !== null &&
```
Comment (ds161833) [inline Client/src/Components/NavigationMenu.tsx]: user && user.sessionExpiration && `<SessionTimeOut />`

File: Client/src/Components/SessionTimeout.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    const { user, setUser } = useContext(UserContext);
+    const [seconds, setSeconds] = React.useState(user.sessionExpiration);
+    const [open, setOpen] = useState<boolean>(false)
+    const history = useHistory()
+
+    useInterval(() => {
+        if (seconds === 1) {
+            redirectToLogin()
+        }
+        else if (seconds < 30) {
```
Comment (ds161833) [inline Client/src/Components/SessionTimeout.tsx]: i dont think 30 seconds is enough for the user to react to being logged out, save their progress and relogin, it should be in range of minutes, like 5 minutes maybe

File: Client/src/Components/SessionTimeout.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            setOpen(true)
+        }
+        setSeconds(seconds - 1)
+    }, 1000)
+
+    const redirectToLogin = async () => {
+        setUser(defaultUserAccountModel)
+        endUserSession()
```
Comment (ds161833) [inline Client/src/Components/SessionTimeout.tsx]: image
setUser already Removes the guy from storage, that makes the endUserSession() only make a call to signout, i think endUserSession can be removed and `await callSignOut()` can be directly called here

File: Client/src/Components/SessionTimeout.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        history.push({
+            pathname: loginRoute
+        })
+    }
+
+    return (
+        <div onClick={redirectToLogin}>
```
Comment (ds161833) [inline Client/src/Components/SessionTimeout.tsx]: I thought you were going to do the button?

File: Client/src/Components/UserReview/UserReviewRow.tsx [hunk 5]
```diff
@@ -1,16 +1,30 @@
...
 }
 
+export const CheckRemainingSessiontime = () => {
+  const { user, setUser } = useContext(UserContext)
+
+  return (
+    <>
+      {user.sessionExpiration ?
+        < Button > The answer is {user.sessionExpiration}.</Button> :
+        < div > Login here.</div>
```
Comment (ds161833) [inline Client/src/Components/UserReview/UserReviewRow.tsx]: what are these white spaces around tags

File: Client/src/Components/UserReview/UserReviewRow.tsx [hunk 6]
```diff
@@ -40,6 +54,8 @@ export const UserReviewRow = (props: IProps) => {
     reload()
   }
 
+
```
Comment (ds161833) [inline Client/src/Components/UserReview/UserReviewRow.tsx]: two extra blanks

File: Server/src/services/authenticationService.ts [hunk 7]
```diff
@@ -177,13 +177,15 @@ export class AuthenticationService {
...
     let token: string;
+    let tokenInformation;
     let jwtExpiry: number = 3000;
     const jwtAccessKey = process.env.ACCESS_SECRET_KEY;
 
     token = await jwt.sign({ jwtPayload }, jwtAccessKey, {
       expiresIn: jwtExpiry
     })
-    return token;
+    tokenInformation = [token, jwtExpiry]
```
Comment (ds161833) [inline Server/src/services/authenticationService.ts:187]: nice

File: Client/src/App.tsx [hunk 8]
```diff
@@ -2,34 +2,35 @@ import React, { useState } from 'react'
 import { getUserFromStorage, putUserInStorage } from './Common/Storage'
 
 import { IUserAccountModel } from './Models/Authentication/IUserAccountModel'
-import NavigationMenu from './Components/NavigationMenu'
+import NavigationMenu from './Components/NavigationMenu';
```
Comment (laithAwad) [inline Client/src/App.tsx]: you don't need ";" at the end, just to keep up with the syntax we are using

File: Client/src/Common/Storage.tsx [hunk 9]
```diff
@@ -1,12 +1,16 @@
-import { IUserAccountModel } from "../Models/Authentication/IUserAccountModel"
+import { IUserAccountModel, defaultUserAccountModel } from '../Models/Authentication/IUserAccountModel';
```
Comment (laithAwad) [inline Client/src/Common/Storage.tsx]: same thing no need for the " ; "

File: Client/src/Components/SessionTimeout.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+import { Alert } from '@material-ui/lab';
+import { Snackbar } from '@material-ui/core';
+import { UserContext } from "../App";
+import { defaultUserAccountModel } from '../Models/Authentication/IUserAccountModel';
+
+import useInterval from 'react-useinterval';
+import { endUserSession } from "../Common/GenericHelpers";
+import { useHistory } from "react-router";
+import { loginRoute } from "../Common/Consts/Routes";
```
Comment (laithAwad) [inline Client/src/Components/SessionTimeout.tsx]: same thing for all these imports, no need for the " ; "

File: Client/src/Components/SessionTimeout.tsx (hunk 2 above)
Comment (laithAwad) [reply Client/src/Components/SessionTimeout.tsx]: I agree, there is no sensitive information so the session can be 5 minutes

File: Client/src/Components/SessionTimeout.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        history.push({
+            pathname: loginRoute
+        })
+    }
+
+    return (
+        <div onClick={redirectToLogin}>
+            <Snackbar open={open} onClose={() => setOpen(false)}>
+                <Alert onClose={() => setOpen(false)} severity="warning">
+                    Session ending soon.... {seconds} Time left. Click here to login again :)
```
Comment (laithAwad) [inline Client/src/Components/SessionTimeout.tsx]: The stakeholder will probably request the removal of " :) " saying it is a machine and not a human we are interacting with, similar to his request of removing "hello" when signing in before

**Review by laithAwad — COMMENTED** [review]
Great work, some minor syntax changes after you review Dima's comments

File: Client/src/Components/SessionTimeout.tsx (hunk 2 above)
Comment (GD-Dheer) [reply Client/src/Components/SessionTimeout.tsx]: Done

File: Client/src/Components/SessionTimeout.tsx (hunk 4 above)
Comment (GD-Dheer) [reply Client/src/Components/SessionTimeout.tsx]: I am not

File: Client/src/Components/UserReview/UserReviewRow.tsx (hunk 5 above)
Comment (GD-Dheer) [reply Client/src/Components/UserReview/UserReviewRow.tsx]: This code is not needed

File: Client/src/Components/SessionTimeout.tsx (hunk 3 above)
Comment (GD-Dheer) [reply Client/src/Components/SessionTimeout.tsx]: Great

**Review by laithAwad — APPROVED** [review]

**Review by ds161833 — APPROVED** [review]
👍

**Pull request merged**